### DISCUSSION TITLE
Unexpected /var/lib/docker size grow

### DISCUSSION THREAD

**Discussion by gray380** [DependencyTrack/dependency-track#2112] [Q&A]
Hello!

Could you explain what data are stored here (and why here, I suppose it should be kept in the /data)
```
/root/.dependency-track/nist
```
and what is the estimated disk capacity utilization for these data?

### Environment:
  
  - Dependency-Track Version:  v4.5.0
  - Distribution: [ Docker ]
  - BOM Format & Version:
  - Database Server: [  MSSQL  ]

Best regards
Serhiy.

**Comment by nscuro** [comment]
See https://docs.dependencytrack.org/getting-started/data-directory/

The data is stored at `/data/.dependency-track/nist` in the container. It should not reside under `/root`.
The `nist` directory is about ~1GB in my local test instance.

**Comment by gray380** [comment]
Thanks for the link.
According to it (as far as I understood):

Dependency-Track uses ~/.dependency-track on UNIX/Linux systems ...
`~/` means user's hone directory

This directory, referred to as the data directory...
`data` is reference, not the real `/data/` directory.

So the question is how to move `.dependency-track` from the `~/` to the `/data`?

UPDATE:
Seems that `/data` is only intended for some templates
```
    volumes:
    # Optional volume mount to override default notification publisher templates
    # - "/host/path/to/template/base/dir:/data/templates" 
      - 'dependency-track:/data'
```

**Reply by nscuro** [reply]
The container is running as non-root user per default, for whom the home directory is set to `/data`, see https://github.com/DependencyTrack/dependency-track/blob/master/src/main/docker/Dockerfile#L37-L44

Did you change the user the container is running as in your `docker-compose.yml`?

> Seems that /data is only intended for some templates

No. The sub directory `templates` of `/data` is, but not the whole `/data` directory.

**Comment by gray380** [comment]
To be honest I didn't modify the Dockerfile and used it as-is:
```
services:
  dtrack-apiserver:
    image: registry.api.bank.lan:8083/common/common.dtrack-apiserver:4.5.0
```
and didn't add any `user: "uid:gid" to the docker-compose.yml.

So I have the following situation:
```
$ docker-compose exec -it dtrack-apiserver id
uid=0(root) gid=0(root) groups=0(root)
```
I assume that the reason is docker's uid mismatching, in my case:
```
$ id
uid=1001(docker) gid=1001(docker)
```
and should be:
```
Ensure UID 1000 & GID 1000 own all the needed directories
```

**Reply by nscuro** [reply]
> registry.api.bank.lan:8083/common/common.dtrack-apiserver

Could it be that the image has been modified by your organization?

**Reply by gray380** [reply]
no, it's only downloaded to the local registry

**Reply by gray380** [reply]
sorry, my fault
yes, it has been modified...

**Comment by gray380** [comment]
I've added to the end of Dockerfile:
USER 1001:1001

and now have the:
```
$ docker-compose exec -it dtrack-apiserver id
uid=1001 gid=1001 groups=1001
```

but the following errors have appeared:
```
ERROR [KeyManager] An error occurred saving newly generated keypair
ERROR [KeyManager] /opt/owasp/dependency-track/?/.dependency-track/keys/public.key
ERROR [KeyManager] An error occurred saving newly generated secret key
ERROR [KeyManager] /opt/owasp/dependency-track/?/.dependency-track/keys/secret.key
```

**Reply by nscuro** [reply]
I'd recommend to use the UID:GID combo `1000:1000` as it's used in the original `Dockerfile`:

https://github.com/DependencyTrack/dependency-track/blob/45868a937f9e5902bee630a2a8c627e459b773d1/src/main/docker/Dockerfile#L11-L12

Also, you will need to fix the permissions of the existing files, as they've been created by the `root` user before. Something along the lines of:

```shell
docker exec --user 0 -it dtrack-apiserver /bin/bash
chown -R 1000:1000 /data
```

**Reply by gray380** [reply]
Thanx a lot for your assistance!
